svf/include/Util/SVFBugReport.h

```diff
@@ -0,0 +1,323 @@
+    inline GenericBug(BugType bugType, EventStack bugEventStack): bugType(bugType), bugEventStack(bugEventStack){ };
+
+    virtual ~GenericBug() = default;
+
+    /// returns bug type
+    inline BugType getBugType() const { return bugType; }
+    /// returns bug location as json format string
+    const std::string getLoc() const;
+    /// return bug source function name
+    const std::string getFuncName() const;
+
+    inline const EventStack& getEventStack() const { return bugEventStack; }
+
+    virtual cJSON *getBugDescription() = 0;
+    virtual void printBugToTerminal() = 0;
+};
+
+class BufferOverflowBug: public GenericBug{
+protected:
+    long long allocLowerBound, allocUpperBound, accessLowerBound, accessUpperBound;
+
+public:
+    inline BufferOverflowBug(GenericBug::BugType bugType, EventStack eventStack,
+                             long long allocLowerBound, long long allocUpperBound,
+                             long long accessLowerBound, long long accessUpperBound):
+          GenericBug(bugType, eventStack), allocLowerBound(allocLowerBound),
+          allocUpperBound(allocUpperBound), accessLowerBound(accessLowerBound),
+          accessUpperBound(accessUpperBound){ }
+
+    cJSON *getBugDescription();
+    void printBugToTerminal();
+
+    /// ClassOf
+    static inline bool classof(const GenericBug *bug)
+    {
+        return bug->getBugType() == GenericBug::PARTIALBUFOVERFLOW || bug->getBugType() == GenericBug::FULLBUFOVERFLOW;
+    }
+};
+
+class FullBufferOverflowBug: public BufferOverflowBug{
+public:
+    inline FullBufferOverflowBug(EventStack eventStack,
+                             long long allocLowerBound, long long allocUpperBound,
+                             long long accessLowerBound, long long accessUpperBound):
+          BufferOverflowBug(GenericBug::FULLBUFOVERFLOW, eventStack, allocLowerBound,
+                            allocUpperBound, accessLowerBound, accessUpperBound){ }
+
+    /// ClassOf
+    static inline bool classof(const GenericBug *bug)
+    {
+        return bug->getBugType() == GenericBug::FULLBUFOVERFLOW;
+    }
+};
+
+class PartialBufferOverflowBug: public BufferOverflowBug{
+protected:
+    long long allocLowerBound, allocUpperBound, accessLowerBound, accessUpperBound;
+
+public:
+    inline PartialBufferOverflowBug( EventStack eventStack,
+                                 long long allocLowerBound, long long allocUpperBound,
+                                 long long accessLowerBound, long long accessUpperBound):
+          BufferOverflowBug(GenericBug::PARTIALBUFOVERFLOW, eventStack, allocLowerBound,
+                            allocUpperBound, accessLowerBound, accessUpperBound){ }
+
+    /// ClassOf
+    static inline bool classof(const GenericBug *bug)
+    {
+        return bug->getBugType() == GenericBug::PARTIALBUFOVERFLOW;
+    }
+};
+
+class NeverFreeBug : public GenericBug{
+public:
+    NeverFreeBug(EventStack bugEventStack):
```

Comment:
Use “const EventStack& “ to avoid copy construction

Reply:
> I'm afraid that copy construction is necessary here because the EventStack created by detector is a function local variable, so it had better to be copied and saved.

It should be const reference because it will do the copy in GeneralBug’s constructor.